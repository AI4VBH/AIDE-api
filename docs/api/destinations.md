<!--
  ~ Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~ http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->

# Destinations

The `destinations` endpoint provides the following APIs to configure the export destinations for workflow results.

## `GET /destinations`

Returns a list of destinations.

### Parameters

N/A

### Responses

| Code | Description |
|------|-------------|
| `200` | Destinations retrieved successfully. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request POST 'http://localhost:5000/destinations'
```

### Example Response

```json
[
  {
    "port": 104,
    "name": "USEAST",
    "aeTitle": "PACSUSEAST",
    "hostIp": "10.20.3.4"
  }
]
```

---

## `POST /destinations`

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Body | | See the `IDestination` interface for details. |

### Responses

| Code | Description |
|------|-------------|
| `201` | Destination created successfully. |
| `400` | Validation error. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request POST 'http://localhost:5000/destinations' \
    --header 'Content-Type: application/json' \
    --data-raw '{
        "name": "USEAST",
        "hostIp": "10.20.3.4",
        "port": 104,
        "aeTitle": "PACSUSEAST"
    }'
```

### Example Response

```json
{
  "port": 104,
  "name": "USEAST",
  "aeTitle": "PACSUSEAST",
  "hostIp": "10.20.3.4"
}
```

---

## `PUT /destinations/:name`

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:name` | `string` corresponding to the destination name. |
| Body | | See the `IDestination` interface for details. |


### Responses

| Code | Description |
|------|-------------|
| `200` | DICOM destination updated successfully. |
| `400` | Validation error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with details of the validation errors . |
| `404` | DICOM destination cannot be found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request PUT 'http://localhost:5000/destinations/USEAST' \
    --header 'Content-Type: application/json' \
    --data-raw '{
        "name": "USEAST",
        "hostIp": "10.20.3.4",
        "port": 104,
        "aeTitle": "PACSUSEAST"
    }'
```

### Example Response

```json
{
  "port": 104,
  "name": "USEAST",
  "aeTitle": "PACSUSEAST",
  "hostIp": "10.20.3.4"
}
```

---

## `GET /destinations/echo/:name`

Checks the status of a destination

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:name` | `string` corresponding to the destination name. |

### Responses

| Code | Description |
|------|-------------|
| `200` | Echo successful. |
| `404` | Destination not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |
| `502` | C-ECHO failure. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/destinations/echo/USEAST'
```