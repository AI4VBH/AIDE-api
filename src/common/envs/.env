PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
