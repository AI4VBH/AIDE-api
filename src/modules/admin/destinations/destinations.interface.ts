export interface IDestination {
  port: number;
  name: string;
  aeTitle: string;
  hostIp: string;
}
