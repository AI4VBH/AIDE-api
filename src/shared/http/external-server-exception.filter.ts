import {
  ArgumentsHost,
  Catch,
  ExceptionFilter,
  HttpException,
  HttpStatus,
  Logger,
} from '@nestjs/common';
import { Response } from 'express';
import { AxiosError } from 'axios';
import { WorkflowServiceException } from 'modules/workflows/workflow.service.exceptions';
import { MinoiClientException } from 'shared/minio/minio-client';
import {
  ExecutionsServiceException,
  ExecutionsServiceExceptionType,
} from 'modules/admin/executions/executions.service.exceptions';

@Catch(Error)
export default class ExternalServerExceptionFilter implements ExceptionFilter {
  catch(exception: Error, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const logger = new Logger('ExceptionFilter');

    logger.error(exception, JSON.stringify(exception, null, 2));

    if (exception instanceof HttpException) {
      const status = exception.getStatus();

      return response.status(status).json({
        statusCode: status,
        message: exception.message,
      });
    }

    if (Object.keys(exception).includes('config')) {
      const { message, status } = (
        exception as AxiosError
      ).toJSON() as ResponseException;

      if (
        status === HttpStatus.REQUEST_TIMEOUT ||
        status >= HttpStatus.INTERNAL_SERVER_ERROR
      ) {
        return response.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
          statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
          message: 'An issue occurred with the MONAI service',
        });
      }

      if (
        status === HttpStatus.BAD_REQUEST ||
        status === HttpStatus.NOT_FOUND
      ) {
        return response
          .status(status)
          .json((exception as AxiosError).response.data);
      }

      return response.status(status ?? HttpStatus.INTERNAL_SERVER_ERROR).json({
        statusCode: status ?? HttpStatus.INTERNAL_SERVER_ERROR,
        message,
      });
    }

    if (exception instanceof WorkflowServiceException) {
      return response.status(HttpStatus.BAD_REQUEST).json({
        statusCode: HttpStatus.BAD_REQUEST,
        message: exception.message,
      });
    }

    if (exception instanceof MinoiClientException) {
      return response.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
        statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
        message: 'An issue occurred with the MINIO service',
      });
    }

    if (exception instanceof ExecutionsServiceException) {
      if (exception.type === ExecutionsServiceExceptionType.MISSING_TASK) {
        return response.status(HttpStatus.NOT_FOUND).json({
          statusCode: HttpStatus.NOT_FOUND,
          message: exception.message,
        });
      }
    }

    response.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
      statusCode: 500,
      message: 'Internal Server Error',
    });
  }
}

type ResponseException = {
  status: number;
  message: string;
};
