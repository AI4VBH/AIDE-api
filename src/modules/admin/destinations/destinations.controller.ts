import {
  BadRequestException,
  Body,
  Controller,
  Get,
  Inject,
  Param,
  Post,
  Put,
  UseFilters,
} from '@nestjs/common';
import { Roles } from 'nest-keycloak-connect';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { IDestination } from './destinations.interface';
import { DestinationsService } from './destinations.service';

@Controller('destinations')
@Roles({ roles: ['realm:admin'] })
@UseFilters(ExternalServerExceptionFilter)
export class DestinationsController {
  @Inject(DestinationsService)
  private readonly service: DestinationsService;

  @Get()
  getDestinations() {
    return this.service.getDestinations();
  }

  @Post()
  registerDestination(@Body() destination: IDestination) {
    this.validateDestination(destination);

    return this.service.registerDestination(destination);
  }

  @Put(':name')
  updateDestination(@Body() destination: IDestination) {
    this.validateDestination(destination);

    return this.service.updateDestination(destination);
  }

  @Get('echo/:name')
  echoDestination(@Param('name') name) {
    return this.service.echoDestination(name);
  }

  private validateDestination(destination: IDestination) {
    if (!destination) {
      throw new BadRequestException('destination object cannot be empty');
    }

    const missingProps: string[] = [];

    ['aeTitle', 'hostIp', 'name', 'port'].forEach(
      (key) => !destination[key] && missingProps.push(key),
    );

    if (missingProps.length !== 0) {
      throw new BadRequestException(
        `The following properties are missing from the destinations object: ${missingProps.join(
          ', ',
        )}`,
      );
    }
  }
}
