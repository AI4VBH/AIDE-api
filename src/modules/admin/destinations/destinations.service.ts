/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { lastValueFrom } from 'rxjs';
import { IDestination } from './destinations.interface';

@Injectable()
export class DestinationsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(ConfigService)
  private readonly configService: ConfigService;

  async deleteDestination(name: string) {
    const baseURL = this.configService.get<string>('MIG_API_HOST');

    const response = await lastValueFrom(
      this.httpService.delete<IDestination>(`/config/destination/${name}`, {
        baseURL,
      }),
    );

    return response.data;
  }

  async getDestinations(): Promise<IDestination[]> {
    const baseURL = this.configService.get<string>('MIG_API_HOST');

    const response = await lastValueFrom(
      this.httpService.get<IDestination[]>('/config/destination', {
        baseURL,
      }),
    );

    return response.data;
  }

  async registerDestination(destination: IDestination) {
    const baseURL = this.configService.get<string>('MIG_API_HOST');

    const result = await lastValueFrom(
      this.httpService.post('/config/destination', destination, {
        baseURL,
      }),
    );

    return result.data;
  }

  async updateDestination(destination: IDestination) {
    const baseURL = this.configService.get<string>('MIG_API_HOST');

    const result = await lastValueFrom(
      this.httpService.put('/config/destination', destination, {
        baseURL,
      }),
    );

    return result.data;
  }

  async echoDestination(name: string) {
    const baseURL = this.configService.get<string>('MIG_API_HOST');

    const result = await lastValueFrom(
      this.httpService.get(`/config/destination/cecho/${name}`, {
        baseURL,
      }),
    );

    if (result.status !== 200) {
      throw new Error('Echo destination request failed');
    }

    return result.data;
  }
}
