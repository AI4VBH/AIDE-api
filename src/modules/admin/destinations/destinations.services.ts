import { HttpService } from '@nestjs/axios';
import { Inject } from '@nestjs/common';
import { IDestination } from './destinations.interface';

export class DestinationsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  getDestinations() {
    return null;
  }

  registerDestination(destination: IDestination) {
    return null;
  }
}
