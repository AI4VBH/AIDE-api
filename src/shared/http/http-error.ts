import { HttpStatus } from "@nestjs/common/enums";

export class HttpError extends Error {
    constructor(statusCode: HttpStatus, message: string) {
      super(message);
  
      this.statusCode = statusCode;
    }
  
  readonly statusCode: HttpStatus;
  }
  
  