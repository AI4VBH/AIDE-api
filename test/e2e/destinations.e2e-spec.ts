import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import { DestinationsController } from 'modules/admin/destinations/destinations.controller';
import { DestinationsService } from 'modules/admin/destinations/destinations.service';
import DestinationsMock from '../test_data/mocks/destinations/destinations-index';

const server = setupServer();
const testMonaiBasePath = 'https://localhost:7337';
const testMigBasePath = 'https://localhost:7338';

describe('/destinations Integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          load: [
            () => ({
              MONAI_API_HOST: testMonaiBasePath,
              MIG_API_HOST: testMigBasePath,
            }),
          ],
        }),
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [DestinationsController],
      providers: [DestinationsService],
    }).compile();

    app = moduleFixture.createNestApplication();

    await app.init();
  });

  it('(GET) /destinations no destinations', async () => {
    server.use(
      rest.get(`${testMigBasePath}/config/destination`, (_req, res, ctx) => {
        return res(ctx.json([]));
      }),
    );
    const response = await request(app.getHttpServer()).get('/destinations');
    expect(response.body).toMatchObject([]);
    expect(response.statusCode).toBe(200);
  });

  it.each([
    DestinationsMock.destinationsList1,
    DestinationsMock.destinationsList2,
  ])('(GET) /destinations with destinations', async (...destinations) => {
    server.use(
      rest.get(`${testMigBasePath}/config/destination`, (_req, res, ctx) => {
        return res(ctx.json(destinations));
      }),
    );
    const response = await request(app.getHttpServer()).get('/destinations');
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(200);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /destinations correct status when MIG gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(`${testMigBasePath}/config/destination`, (_req, res, ctx) => {
          return res(ctx.status(code));
        }),
      );
      const response = await request(app.getHttpServer()).get('/destinations');
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it('(POST) /destinations/ with body', async () => {
    server.use(
      rest.post(`${testMigBasePath}/config/destination`, (req, res, ctx) => {
        return res(
          ctx.status(201),
          ctx.json(DestinationsMock.destinationsObject1),
        );
      }),
    );
    const response = await request(app.getHttpServer())
      .post('/destinations')
      .send(DestinationsMock.destinationsObject1);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(201);
  });

  it('(DELETE) /destinations/ with url param', async () => {
    server.use(
      rest.delete(
        `${testMigBasePath}/config/destination/Lillie`,
        (req, res, ctx) => {
          return res(
            ctx.status(200),
            ctx.json(DestinationsMock.destinationsObject1),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).delete(
      '/destinations/Lillie',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(200);
  });

  it.each([
    DestinationsMock.destinationsObjectMissingAeTitle,
    DestinationsMock.destinationsObjectMissingName,
    DestinationsMock.destinationsObjectMissingPort,
    DestinationsMock.destinationsMissingHostIp,
    DestinationsMock.destinationsObjectEmpty,
  ])('(POST) /destinations/ Invalid body', async (destinationBody) => {
    server.use(
      rest.post(`${testMigBasePath}/config/destination`, (req, res, ctx) => {
        return res(ctx.status(400), ctx.json(destinationBody));
      }),
    );
    const response = await request(app.getHttpServer())
      .post(`/destinations`)
      .send(destinationBody);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(POST) /destination/:name correct status when MIG gives general error with code %s',
    async (code) => {
      server.use(
        rest.post(`${testMigBasePath}/config/destination`, (_req, res, ctx) => {
          return res(ctx.status(code));
        }),
      );
      const response = await request(app.getHttpServer())
        .post(`/destinations`)
        .send(DestinationsMock.destinationsObject1);
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it('(PUT) /destinations/:name with body', async () => {
    server.use(
      rest.put(`${testMigBasePath}/config/destination`, (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json(DestinationsMock.destinationsObject1),
        );
      }),
    );
    const response = await request(app.getHttpServer())
      .put('/destinations/ORTHANC')
      .send(DestinationsMock.destinationsObject1);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(200);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(PUT) /destinations/:name correct status when MIG gives general error with code %s',
    async (code) => {
      server.use(
        rest.put(`${testMigBasePath}/config/destination`, (_req, res, ctx) => {
          return res(ctx.status(code));
        }),
      );
      const response = await request(app.getHttpServer())
        .put('/destinations/ORTHANC')
        .send(DestinationsMock.destinationsObject1);
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it.each([
    DestinationsMock.destinationsObjectMissingAeTitle,
    DestinationsMock.destinationsObjectMissingName,
    DestinationsMock.destinationsObjectMissingPort,
    DestinationsMock.destinationsMissingHostIp,
    DestinationsMock.destinationsObjectEmpty,
  ])('(PUT) /destinations/:name Invalid body', async (destinationBody) => {
    server.use(
      rest.put(`${testMigBasePath}/config/destination`, (req, res, ctx) => {
        return res(ctx.status(400), ctx.json(destinationBody));
      }),
    );
    const response = await request(app.getHttpServer())
      .put(`/destinations/ORTHANC`)
      .send(destinationBody);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it('(PUT) /destinations/:name changed name', async () => {
    server.use(
      rest.put(`${testMigBasePath}/config/destination`, (req, res, ctx) => {
        return res(
          ctx.status(404),
          ctx.json(DestinationsMock.destinationsError404),
        );
      }),
    );
    const response = await request(app.getHttpServer())
      .put('/destinations/ORTHANC')
      .send(DestinationsMock.destinationsObject1);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(404);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(PUT) /destination/:name correct status when MIG gives general error with code %s',
    async (code) => {
      server.use(
        rest.put(`${testMigBasePath}/config/destination`, (_req, res, ctx) => {
          return res(ctx.status(code));
        }),
      );
      const response = await request(app.getHttpServer())
        .put(`/destinations/ORTHANC`)
        .send(DestinationsMock.destinationsObject1);
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it('(GET) /destinations/echo/:name successfully dicom echo a destination', async () => {
    const name = 'ORTHANC';
    server.use(
      rest.get(
        `${testMigBasePath}/config/destination/cecho/${name}`,
        (_req, res, ctx) => {
          return res(ctx.status(200));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      `/destinations/echo/${name}`,
    );
    expect(response.statusCode).toBe(200);
  });

  it('(GET) /destinations/echo/:name dicom echo a missing destination ', async () => {
    const name = 'missing';
    server.use(
      rest.get(
        `${testMigBasePath}/config/destination/cecho/${name}`,
        (_req, res, ctx) => {
          return res(
            ctx.status(404),
            ctx.json(DestinationsMock.destinationsError404),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      `/destinations/echo/${name}`,
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(404);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /destinations correct status when MIG gives general error with code %s',
    async (code) => {
      const name = 'ORTHANC';
      server.use(
        rest.get(
          `${testMigBasePath}/config/destination/cecho/${name}`,
          (_req, res, ctx) => {
            return res(ctx.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        `/destinations/echo/${name}`,
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );
});
