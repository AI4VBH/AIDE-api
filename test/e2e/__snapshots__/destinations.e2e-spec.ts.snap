// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/destinations Integration Tests (DELETE) /destinations/ with url param 1`] = `
{
  "aeTitle": "AE_TITLE",
  "hostIp": "0.1.1.1",
  "name": "SOME_NAME",
  "port": 5000,
}
`;

exports[`/destinations Integration Tests (GET) /destinations with destinations 1`] = `
[
  {
    "aeTitle": "AE_TITLE",
    "hostIp": "0.1.1.1",
    "name": "SOME_NAME",
    "port": 5000,
  },
]
`;

exports[`/destinations Integration Tests (GET) /destinations with destinations 2`] = `
[
  {
    "aeTitle": "AE_TITLE-2",
    "hostIp": "0.1.1.2",
    "name": "SOME_NAME-2",
    "port": 5002,
  },
  {
    "aeTitle": "AE_TITLE-1",
    "hostIp": "0.1.1.1",
    "name": "SOME_NAME-1",
    "port": 5001,
  },
  {
    "aeTitle": "AE_TITLE-3",
    "hostIp": "0.1.1.3",
    "name": "SOME_NAME-3",
    "port": 5003,
  },
]
`;

exports[`/destinations Integration Tests (GET) /destinations/echo/:name dicom echo a missing destination  1`] = `
{
  "status": 404,
  "title": "Not Found",
  "traceId": "00-75faa21c19aa287947659f52b1a9f797-6535ba8afea43179-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/destinations Integration Tests (POST) /destinations/ Invalid body 1`] = `
{
  "message": "The following properties are missing from the destinations object: aeTitle",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (POST) /destinations/ Invalid body 2`] = `
{
  "message": "The following properties are missing from the destinations object: name",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (POST) /destinations/ Invalid body 3`] = `
{
  "message": "The following properties are missing from the destinations object: port",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (POST) /destinations/ Invalid body 4`] = `
{
  "message": "The following properties are missing from the destinations object: hostIp",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (POST) /destinations/ Invalid body 5`] = `
{
  "message": "The following properties are missing from the destinations object: aeTitle, hostIp, name, port",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (POST) /destinations/ with body 1`] = `
{
  "aeTitle": "AE_TITLE",
  "hostIp": "0.1.1.1",
  "name": "SOME_NAME",
  "port": 5000,
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name Invalid body 1`] = `
{
  "message": "The following properties are missing from the destinations object: aeTitle",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name Invalid body 2`] = `
{
  "message": "The following properties are missing from the destinations object: name",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name Invalid body 3`] = `
{
  "message": "The following properties are missing from the destinations object: port",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name Invalid body 4`] = `
{
  "message": "The following properties are missing from the destinations object: hostIp",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name Invalid body 5`] = `
{
  "message": "The following properties are missing from the destinations object: aeTitle, hostIp, name, port",
  "statusCode": 400,
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name changed name 1`] = `
{
  "status": 404,
  "title": "Not Found",
  "traceId": "00-75faa21c19aa287947659f52b1a9f797-6535ba8afea43179-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/destinations Integration Tests (PUT) /destinations/:name with body 1`] = `
{
  "aeTitle": "AE_TITLE",
  "hostIp": "0.1.1.1",
  "name": "SOME_NAME",
  "port": 5000,
}
`;
